### PR TITLE
feat: improve firewall resources management

### LINKED ISSUE #111: Support Firewall resource selectors
##### SUMMARY
Support apply_to parameter in firewalls to configure firewall resource selectors.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
hetzner.hcloud.hcloud_firewall

##### ADDITIONAL INFORMATION
Right now firewalls cannot be specified using apply_to. This makes them more or less unsuitable if utilizing labels.
Specifying each server manually through hcloud_server_module seems counter unintuitive. Especially for more complex setups.
Note the official api does not support lookup by server names and has a different apply_to syntax. However i think this is more likely to be idomatic in ansible.
Implementations therefore need to resolve the id of the servers automatically.

Otherwise implementation should be more or less straight forward by calling https://api.hetzner.cloud/v1/firewalls , https://api.hetzner.cloud/v1/firewalls/{id}/actions/apply_to_resources and https://api.hetzner.cloud/v1/firewalls/{id}/actions/remove_from_resources
respectively to create or update or delete unused resources.

```yaml
   - name: Create main firewall
      'hetzner.hcloud.hcloud_firewall':
        name: main_fw
        apply_to:
            label_selectors:
                - k
            servers:
                - my-server-name
        rules:
          - direction: in
            protocol: tcp
            port: 22
            source_ips:
              - 0.0.0.0/0
              - ::/0
            description: Allows SSH access
        state: present

```

### PR DESCRIPTION
**Pull request by jooola**
##### SUMMARY

  - firewall - Return resources the firewall is `applied_to`.
  - firewall_info - Add new `firewall_info` module to gather firewalls info.
  - firewall_resource - Add new `firewall_resource` module to manage firewalls resources.

Fixes #111

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

firewall
firewall_info
firewall_resource

### CODE REVIEW

File: tests/integration/targets/firewall_resource/tasks/test.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  hetzner.hcloud.firewall_resource:
+    firewall: "{{ hcloud_firewall_name }}"
+    servers: ["{{ hcloud_server_name }}"]
+  register: result
+- name: Verify create idempotency
+  ansible.builtin.assert:
+    that:
+      - result is not changed
+
+- name: Test update
```
Comment (LarsKumbier) [inline tests/integration/targets/firewall_resource/tasks/test.yml:51]: Missing: the result of an update to a label_selector should remove the server attribute from the firewall (I guess)?

Comment (apricote) [reply tests/integration/targets/firewall_resource/tasks/test.yml:51]: The current behavior of the ansible module matches the API. You can add and remove resources independently.

(Not an ansible expert) This seems to match the general Ansible behavior, only the fields you specify are actually updated, there is no expectation that you specify the full configuration in every module invocation. Ie. you do not specify all installed OS packages when you want to install a new one.

I agree that magically removing the other attribute would not match my expectation.

That said, is there a way to remove all current attachements/resources from a firewall? Probably through the `hcloud_firewall_info` resource and `state: absent`, right?

Comment (jooola) [reply tests/integration/targets/firewall_resource/tasks/test.yml:51]: There is no easy way to remove all the attached resources, using  `hcloud_firewall_info` is an option, but it is indeed suboptimal.

Maybe having an extra firewall_resource state `clean` could be the solution to this?

File: plugins/modules/firewall_info.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                for item in resource.applied_to_resources
+            ]
+        return result
+
+    def get_firewalls(self):
+        try:
+            if self.module.params.get("id") is not None:
+                self.hcloud_firewall_info = [self.client.firewalls.get_by_id(self.module.params.get("id"))]
+            elif self.module.params.get("name") is not None:
+                self.hcloud_firewall_info = [self.client.firewalls.get_by_name(self.module.params.get("name"))]
```
[the remark this answers is not in the document]
Comment (jooola) [reply plugins/modules/firewall_info.py:211]: Yes, I don't want to diverge from the other module that implemented like this. The _prepare_result function loops over the self.hcloud_firewall_info and filter any None values.

File: tests/integration/targets/firewall_resource/tasks/test.yml (hunk 1 above)
Comment (jooola) [reply tests/integration/targets/firewall_resource/tasks/test.yml:51]: I prefer to implement such feature in a future PR, this one is already quite large.

File: plugins/modules/firewall_resource.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+# Copyright: (c) 2019, Hetzner Cloud GmbH <info@hetzner-cloud.de>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+
+from __future__ import annotations
+
+DOCUMENTATION = """
+---
+module: firewall_resource
```
[the remark this answers is not in the document]
Comment (jooola) [reply plugins/modules/firewall_resource.py:11]: I'd prefer not to diverge from the frontend too, which already uses `resources`.

Maybe the terminology can be mentioned in our [docs](docs.hetzner.cloud], "what is a resource" for example?

[the remark this answers is not in the document]
Comment (apricote) [reply plugins/modules/firewall_resource.py:11]: Works for me! Naming things consistently between API, Cloud Console and Integrations is something we should try to improve in the future.

**Review by apricote — APPROVED** [review]

**Pull request merged**